fish-rust/src/curses.rs

```diff
@@ -0,0 +1,261 @@
+//!
+//! In addition to exposing the C library ffi calls, we also shim around some functionality that's
+//! only made available via the the ncurses headers to C code via macro magic, such as polyfilling
+//! missing capability strings to shoe-in missing support for certain terminal sequences.
+//!
+//! This is intentionally very bare bones and only implements the subset of curses functionality
+//! used by fish
+
+use self::sys::*;
+use crate::supercow::ArcCow;
+use once_cell::sync::Lazy;
+use std::ffi::{CStr, CString};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+
+// This is `static mut` because a `Lazy` can only be init once but we may need to re-init the Term
+// value after a second+ call to curses::setup() in response to environment changes.
+pub static mut TERM: Lazy<Term> = Lazy::new(|| panic!("TERM not yet initialized!"));
+
+/// Returns a mutable reference to the global [`Term`] singleton.
+///
+/// Panics on deref if [`curses::setup()`](self::setup()) hasn't been called successfully.
+pub fn term() -> &'static Term {
+    unsafe { &TERM }
+}
+
+/// Private module exposing system curses ffi.
+mod sys {
+    pub const OK: i32 = 0;
+    pub const ERR: i32 = -1;
+
+    extern "C" {
+        /// The ncurses `cur_term` TERMINAL pointer.
+        pub static mut cur_term: *const core::ffi::c_void;
+
+        /// setupterm(3) is a low-level call to begin doing any sort of `term.h`/`curses.h` work.
+        /// It's called internally by ncurses's `initscr()` and `newterm()`, but the C++ code called
+        /// it directly from [`initialize_curses_using_fallbacks()`].
+        pub fn setupterm(
+            term: *const libc::c_char,
+            filedes: libc::c_int,
+            errret: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        /// Frees the `cur_term` TERMINAL  pointer.
+        pub fn del_curterm(term: *const core::ffi::c_void);
+
+        /// Checks for the presence of a termcap flag identified by the first two characters of
+        /// `id`. The C function returns an integer, but we just reinterpret that as a bool.
+        pub fn tgetflag(id: *const libc::c_char) -> bool;
+
+        /// Checks for the presence and value of a number capability in the termcap/termconf
+        /// database. A return value of `-1` indicates not found.
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+// String capabilities
+pub const ENTER_ITALICS_MODE: StringCap = StringCap::new("ZH");
+pub const EXIT_ITALICS_MODE: StringCap = StringCap::new("ZR");
+pub const ENTER_DIM_MODE: StringCap = StringCap::new("mh");
+
+// Number capabilities
+pub const MAX_COLORS: Number = Number::new("Co");
+
+// Flag capabilities
+pub const EAT_NEWLINE_GLITCH: Flag = Flag::new("xn");
+
+pub struct Term {
```

Comment:
See [my previous comment](https://github.com/fish-shell/fish-shell/pull/9786#discussion_r1193217691) - there are no mutexes flying around here, the mutex is restricted only to the interior of the `get()`/`set()` call and never outlives those. 

>  An "override" would just be overwriting any values loaded from setupterm.

The devil hides behind that "just" - we can't (safely) overwrite a value without an &mut reference to the cached list of strings. We can refactor the code later to have `env_dispatch` both initialize term (getting a mutable `Term` instance and not a reference) and set its overrides with that `mut` reference *then* store the `TERM` singleton so that any subsequent borrows of TERM can be `&Term` and we'd never need `&mut Term`. But the way the calling code is currently structured, we arbitrarily need `&mut Term` references so we need the mutex (and the `has_overrides` machinery is just to bypass the mutex when there are no overrides - it's solely an optimization).

Reply:
Actually, since I changed from `&str` to `Arc<CString>` for the reasons explained above, I can drop the mutex and change the vec to `Vec<Cell<Arc<CString>>>` - everything stays the same but we don't need a mutex since the clone out of the cell is "free". I can also go back to caching the result of ncurses lookups since there would never be a mutex so we don't need to avoid caching in order to avoid locking.